<a href="https://colab.research.google.com/github/DevTedd/Individual-Project-Week-4-Moringa/blob/main/Moringa_Data_Science_Prep_W4_Independent_Project_2021_09_Ted_Kimani_DataProcessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#  Friday: Independent Project - Week 4


In this week's independent project, you will be working as a data scientist working for an electric car-sharing service company. You have been tasked to process stations data to understand electric car usage over time by solving for the following research question;

# Research Question

Identify the most popular hour of the day for picking up a shared electric car (Bluecar) in the city of Paris over the month of April 2018.
Bonus Questions (Optional)

What is the most popular hour for returning cars?
What station is the most popular?
Overall?
At the most popular picking hour?
What postal code is the most popular for picking up Blue cars? Does the most popular station belong to that postal code?
Overall?
At the most popular picking hour?
Do the results change if you consider Utilib and Utilib 1.4 instead of Blue cars? 
Your final deliverable will be a data report which will comprise the following sections;




1. Business Understanding 
2. Data Understanding 
3. Data Preparation
4. Analysis 
5. Recommendation 
6. Evaluation






# Import dataset and libraries 

In [1]:
#importing pandas library 
import pandas as pd

#importing numpy library
import numpy as np

In [44]:
df = pd.read_csv('/content/Autolib_dataset.csv')


**Previewing the dataset**

In [7]:
df.head(10)

Address  Cars  Bluecar counter  ...  day  hour  minute
0                 2 Avenue de Suffren     0                0  ...    8    11      43
1           145 Rue Raymond Losserand     6                6  ...    6     7      24
2    2 Avenue John Fitzgerald Kennedy     3                3  ...    3    20      14
3              51 Rue Eugène Oudiné     3                3  ...    4     4      37
4  6 avenue de la Porte de Champerret     3                3  ...    8    17      23
5                8 Boulevard Voltaire     0                0  ...    6     7       2
6                      37 rue Leblanc     0                0  ...    8    18      20
7         17 Rue des Luaps Prolongée     3                3  ...    4    22      13
8               34 avenue Jean Moulin     1                1  ...    2    22      58
9        41 boulevard de Rochechouart     6                6  ...    4    15       2

[10 rows x 25 columns]

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 25 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Address              5000 non-null   object
 1   Cars                 5000 non-null   int64 
 2   Bluecar counter      5000 non-null   int64 
 3   Utilib counter       5000 non-null   int64 
 4   Utilib 1.4 counter   5000 non-null   int64 
 5   Charge Slots         5000 non-null   int64 
 6   Charging Status      5000 non-null   object
 7   City                 5000 non-null   object
 8   Displayed comment    111 non-null    object
 9   ID                   5000 non-null   object
 10  Kind                 5000 non-null   object
 11  Geo point            5000 non-null   object
 12  Postal code          5000 non-null   int64 
 13  Public name          5000 non-null   object
 14  Rental status        5000 non-null   object
 15  Scheduled at         47 non-null     object
 16  Slots 

In [6]:
df.shape

(5000, 25)

#Data Cleaning 


Starting with validity 

In [45]:
#dropping the columns i wont use, ill make a new dataframe without them
df1 = df

df1.drop(['year','Scheduled at','Displayed comment','Geo point'], axis = 1, inplace = True)
#drop the ones with a broken 
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 21 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Address              5000 non-null   object
 1   Cars                 5000 non-null   int64 
 2   Bluecar counter      5000 non-null   int64 
 3   Utilib counter       5000 non-null   int64 
 4   Utilib 1.4 counter   5000 non-null   int64 
 5   Charge Slots         5000 non-null   int64 
 6   Charging Status      5000 non-null   object
 7   City                 5000 non-null   object
 8   ID                   5000 non-null   object
 9   Kind                 5000 non-null   object
 10  Postal code          5000 non-null   int64 
 11  Public name          5000 non-null   object
 12  Rental status        5000 non-null   object
 13  Slots                5000 non-null   int64 
 14  Station type         5000 non-null   object
 15  Status               5000 non-null   object
 16  Subscr

In [46]:
#drop the ones with a broken rental status as they dont have an resources allocated to them
df1.drop(df1[df1['Rental status'] == 'broken'].index, inplace= True)
df1['Rental status']

0       operational
1       operational
2       operational
3       operational
4       operational
           ...     
4995    operational
4996    operational
4997    operational
4998    operational
4999    operational
Name: Rental status, Length: 4668, dtype: object

In [47]:
#dropping these as they are not required to answer any questions
df1.drop(['Subscription status','month','Charging Status'], axis = 1, inplace= True)
df1.head(5)

Address  Cars  Bluecar counter  ...  day  hour  minute
0                 2 Avenue de Suffren     0                0  ...    8    11      43
1           145 Rue Raymond Losserand     6                6  ...    6     7      24
2    2 Avenue John Fitzgerald Kennedy     3                3  ...    3    20      14
3              51 Rue Eugène Oudiné     3                3  ...    4     4      37
4  6 avenue de la Porte de Champerret     3                3  ...    8    17      23

[5 rows x 18 columns]

In [48]:
df1.drop(['ID'], axis = 1, inplace= True)
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4668 entries, 0 to 4999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Address             4668 non-null   object
 1   Cars                4668 non-null   int64 
 2   Bluecar counter     4668 non-null   int64 
 3   Utilib counter      4668 non-null   int64 
 4   Utilib 1.4 counter  4668 non-null   int64 
 5   Charge Slots        4668 non-null   int64 
 6   City                4668 non-null   object
 7   Kind                4668 non-null   object
 8   Postal code         4668 non-null   int64 
 9   Public name         4668 non-null   object
 10  Rental status       4668 non-null   object
 11  Slots               4668 non-null   int64 
 12  Station type        4668 non-null   object
 13  Status              4668 non-null   object
 14  day                 4668 non-null   int64 
 15  hour                4668 non-null   int64 
 16  minute              4668

In [49]:
df1.drop(['Cars'], axis = 1, inplace= True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4668 entries, 0 to 4999
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Address             4668 non-null   object
 1   Bluecar counter     4668 non-null   int64 
 2   Utilib counter      4668 non-null   int64 
 3   Utilib 1.4 counter  4668 non-null   int64 
 4   Charge Slots        4668 non-null   int64 
 5   City                4668 non-null   object
 6   Kind                4668 non-null   object
 7   Postal code         4668 non-null   int64 
 8   Public name         4668 non-null   object
 9   Rental status       4668 non-null   object
 10  Slots               4668 non-null   int64 
 11  Station type        4668 non-null   object
 12  Status              4668 non-null   object
 13  day                 4668 non-null   int64 
 14  hour                4668 non-null   int64 
 15  minute              4668 non-null   int64 
dtypes: int64(9), object(7)
m

The next step is to determine the 
#Completeness & Uniformity 

In [50]:
#checking for any blank spaces
df1.isnull().any()

Address               False
Bluecar counter       False
Utilib counter        False
Utilib 1.4 counter    False
Charge Slots          False
City                  False
Kind                  False
Postal code           False
Public name           False
Rental status         False
Slots                 False
Station type          False
Status                False
day                   False
hour                  False
minute                False
dtype: bool

In [51]:
#checking for any duplicates
df1.duplicated().any()

False

In [36]:
df1.head(5)

Address  Bluecar counter  ...  hour minute
0                 2 Avenue de Suffren                0  ...    11     43
1           145 Rue Raymond Losserand                6  ...     7     24
3              51 Rue Eugène Oudiné                3  ...     4     37
4  6 avenue de la Porte de Champerret                3  ...    17     23
5                8 Boulevard Voltaire                0  ...     7      2

[5 rows x 13 columns]

In [22]:
#creating a data frame with just blue car information
dfBlue = df1
dfBlue.drop(['Utilib counter','Utilib 1.4 counter'], axis = 1, inplace = True)

In [25]:
dfBlue.head(5)
dfBlue.shape

Address  Bluecar counter  ...  hour minute
0                 2 Avenue de Suffren                0  ...    11     43
1           145 Rue Raymond Losserand                6  ...     7     24
2    2 Avenue John Fitzgerald Kennedy                3  ...    20     14
3              51 Rue Eugène Oudiné                3  ...     4     37
4  6 avenue de la Porte de Champerret                3  ...    17     23

[5 rows x 13 columns]

#Beginning the analysis of the dataset

In [27]:
#the first step is isolating only the rides that ocured in paris in april 2018
dfBlue.drop(dfBlue[dfBlue['City'] != 'Paris'].index, inplace= True)
dfBlue['City']

0       Paris
1       Paris
3       Paris
4       Paris
5       Paris
        ...  
4992    Paris
4994    Paris
4996    Paris
4997    Paris
4999    Paris
Name: City, Length: 2641, dtype: object

In [31]:
#well begin by grouping by the hours in order to see hours where the rides are most popular
hourGroup = dfBlue.groupby('hour').count()
hourGroup

Address  Bluecar counter  Charge Slots  ...  Status  day  minute
hour                                          ...                     
0         102              102           102  ...     102  102     102
1          98               98            98  ...      98   98      98
2         112              112           112  ...     112  112     112
3         112              112           112  ...     112  112     112
4         108              108           108  ...     108  108     108
5         113              113           113  ...     113  113     113
6         119              119           119  ...     119  119     119
7         113              113           113  ...     113  113     113
8         122              122           122  ...     122  122     122
9         115              115           115  ...     115  115     115
10        111              111           111  ...     111  111     111
11         98               98            98  ...      98   98      98
12        113              113           113  ...     113  113     113
13        110              110           110  ...     110  110     110
14         99               99            99  ...      99   99      99
15        108              108           108  ...     108  108     108
16        120              120           120  ...     120  120     120
17        111              111           111  ...     111  111     111
18        107              107           107  ...     107  107     107
19        106              106           106  ...     106  106     106
20        103              103           103  ...     103  103     103
21        132              132           132  ...     132  132     132
22        100              100           100  ...     100  100     100
23        109              109           109  ...     109  109     109

[24 rows x 12 columns]

In [34]:
srthourG = hourGroup.sort_values('Bluecar counter', ascending= False)
srthourG

Address  Bluecar counter  Charge Slots  ...  Status  day  minute
hour                                          ...                     
21        132              132           132  ...     132  132     132
8         122              122           122  ...     122  122     122
16        120              120           120  ...     120  120     120
6         119              119           119  ...     119  119     119
9         115              115           115  ...     115  115     115
12        113              113           113  ...     113  113     113
5         113              113           113  ...     113  113     113
7         113              113           113  ...     113  113     113
2         112              112           112  ...     112  112     112
3         112              112           112  ...     112  112     112
10        111              111           111  ...     111  111     111
17        111              111           111  ...     111  111     111
13        110              110           110  ...     110  110     110
23        109              109           109  ...     109  109     109
15        108              108           108  ...     108  108     108
4         108              108           108  ...     108  108     108
18        107              107           107  ...     107  107     107
19        106              106           106  ...     106  106     106
20        103              103           103  ...     103  103     103
0         102              102           102  ...     102  102     102
22        100              100           100  ...     100  100     100
14         99               99            99  ...      99   99      99
1          98               98            98  ...      98   98      98
11         98               98            98  ...      98   98      98

[24 rows x 12 columns]

There was 132 cars ordered in the 21st(9 pm) hour of all the days in the month of April 2018. This is the busyest of all hour slots

Well now move to the bonus questions

In [41]:
#What is the most popular hour for returning cars?
#The logic behind the code below is that most popular hour to return cars is when the parking slots are all full?
#well first group by slots and the number of cars counter
2ndgroup = dfBlue.groupby(['hour', 'Slots']).count()
#well then sort by the hour and ascending oder
2ndgroup.sort_values('')


Address  Bluecar counter  Charge Slots  ...  Status  day  minute
hour Slots                                          ...                     
0    0           23               23            23  ...      23   23      23
     1           20               20            20  ...      20   20      20
     2           11               11            11  ...      11   11      11
     3           17               17            17  ...      17   17      17
     4           15               15            15  ...      15   15      15
...             ...              ...           ...  ...     ...  ...     ...
23   2           15               15            15  ...      15   15      15
     3           15               15            15  ...      15   15      15
     4           13               13            13  ...      13   13      13
     5           16               16            16  ...      16   16      16
     6            8                8             8  ...       8    8       8

[168 rows x 11 columns]

In [42]:
#What station is the most popular? Overall
df1
PopGroup = dfBlue.groupby(['Postal code','Bluecar counter']).count()
PopGroup.sort_values('Bluecar counter', ascending = False)

Address  Charge Slots  City  ...  day  hour  minute
Postal code Bluecar counter                               ...                   
75011       7                      2             2     2  ...    2     2       2
75003       6                      1             1     1  ...    1     1       1
75020       6                     15            15    15  ...   15    15      15
75019       6                      4             4     4  ...    4     4       4
75018       6                     10            10    10  ...   10    10      10
...                              ...           ...   ...  ...  ...   ...     ...
75015       0                     54            54    54  ...   54    54      54
75016       0                     41            41    41  ...   41    41      41
75017       0                     42            42    42  ...   42    42      42
75002       0                     37            37    37  ...   37    37      37
75001       0                     36            36    36  ...   36    36      36

[140 rows x 11 columns]

The station at postal code 75011 was the most popular

In [57]:
df1.head(5)
dfRed = df1
stationPop = dfRed.groupby(['Public name', 'Bluecar counter','Utilib counter','Utilib 1.4 counter']).count()
stationPop.sort_values('Bluecar counter', ascending = False).head(15)

Address  ...  minute
Public name                            Bluecar counter Utilib counter Utilib 1.4 counter           ...        
Vélizy-Villacoublay/Paul Dautier/2    7               0              0                         2  ...       2
Paris/Renault/2                        7               0              0                         2  ...       2
Versailles/Bonne Aventure/68           7               0              0                         2  ...       2
Colombes/Parc Lagravère/1             7               0              0                         1  ...       1
Sceaux/Franklin Roosevelt/60           7               0              0                         1  ...       1
Nanterre/Lavoisier/14                  7               0              0                         1  ...       1
Aulnay-sous-bois/Isidore Nerat/6       7               0              0                         1  ...       1
Les Lilas/Léon Renault/19             7               0              0                         1  ...       1
Le Chesnay/Pottier/7                   7               0              0                         1  ...       1
Nanterre/Georges Clémenceau/205       7               0              0                         2  ...       2
Drancy/Saint Stenay/29                 7               0              0                         1  ...       1
Limeil-Brévannes/Saint John Perse/20  7               0              0                         1  ...       1
Asnières-sur-Seine/Normandie/22       7               0              0                         1  ...       1
Fontenay-aux-Roses/Division Leclerc/31 7               0              0                         2  ...       2
Le Plessis-Robinson/Galilée/16        7               0              0                         7  ...       7

[15 rows x 12 columns]

Its not possible to tell which station is the most popular as a number of the have the same number of cars utilize them

In [63]:
#Finding if the postal code is found is part of the most popular list 
result = dfRed[dfRed['Postal code'] == 75011]
result.groupby(['Public name']).count()
#theres a large number of stations in this postal code

Address  Bluecar counter  ...  hour  minute
Public name                                                  ...              
Paris/Commandant Lamy/2                  7                7  ...     7       7
Paris/Faidherbe/27                       5                5  ...     5       5
Paris/François de Neufchateau/11        4                4  ...     4       4
Paris/Ledru Rollin/104                   4                4  ...     4       4
Paris/Ledru Rollin/155                   5                5  ...     5       5
Paris/Léon Frot/1                       2                2  ...     2       2
Paris/Parmentier/125                     6                6  ...     6       6
Paris/Parmentier/2                       6                6  ...     6       6
Paris/Parmentier/62                      4                4  ...     4       4
Paris/Philippe Auguste/1                 3                3  ...     3       3
Paris/Philippe Auguste/126              11               11  ...    11      11
Paris/Philippe Auguste/65                5                5  ...     5       5
Paris/Renault/2                          5                5  ...     5       5
Paris/Richard Lenoir/110                 3                3  ...     3       3
Paris/Richard Lenoir/36                  5                5  ...     5       5
Paris/Richard Lenoir/63                  7                7  ...     7       7
Paris/Richard Lenoir/8                   5                5  ...     5       5
Paris/Roquette/162                       4                4  ...     4       4
Paris/Roubo/1                            1                1  ...     1       1
Paris/République/128                    7                7  ...     7       7
Paris/République/36                     5                5  ...     5       5
Paris/République/5                      4                4  ...     4       4
Paris/République/86                     7                7  ...     7       7
Paris/Saint-Antoine/199                  6                6  ...     6       6
Paris/Saint-Maur/157                     4                4  ...     4       4
Paris/Saint-Maur/36                      7                7  ...     7       7
Paris/Taillebourg/2                      9                9  ...     9       9
Paris/Voltaire/106                       6                6  ...     6       6
Paris/Voltaire/182                      11               11  ...    11      11
Paris/Voltaire/236                       8                8  ...     8       8
Paris/Voltaire/8                         7                7  ...     7       7

[31 rows x 15 columns]